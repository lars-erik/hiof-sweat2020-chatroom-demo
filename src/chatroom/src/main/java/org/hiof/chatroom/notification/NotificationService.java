package org.hiof.chatroom.notification;

import org.hiof.chatroom.core.ChatMessage;

public interface NotificationService {
    void notifyNewMessage(ChatMessage message);
}
