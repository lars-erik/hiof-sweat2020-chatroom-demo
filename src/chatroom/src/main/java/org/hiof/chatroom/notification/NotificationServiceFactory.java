package org.hiof.chatroom.notification;

public abstract class NotificationServiceFactory {
    public static NotificationServiceFactory instance;

    public abstract NotificationService getService();
}
