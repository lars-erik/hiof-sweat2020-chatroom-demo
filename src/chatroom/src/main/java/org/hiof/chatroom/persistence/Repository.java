package org.hiof.chatroom.persistence;

import org.hiof.chatroom.core.ChatMessage;

import java.util.stream.Stream;

public interface Repository<T> {
    void add(T message);
    Stream<T> all();
    Stream<T> query(Query query) throws Exception;
    T get(String id);
}
