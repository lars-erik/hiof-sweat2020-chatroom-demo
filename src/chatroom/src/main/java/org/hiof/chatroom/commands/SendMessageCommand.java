package org.hiof.chatroom.commands;

import org.hiof.chatroom.core.*;
import org.hiof.chatroom.notification.*;
import org.hiof.chatroom.persistence.*;

import java.util.UUID;

public class SendMessageCommand {
    public String user;
    public String message;

    public void execute() throws Exception {
        UnitOfWork uow = PersistenceFactory.instance.createUnitOfWork();
        ChatMessageRepository repo = PersistenceFactory.instance.createChatMessageRepository(uow);

        ChatMessage msg = new ChatMessage();
        msg.setId(UUID.randomUUID());
        msg.setUser(user);
        msg.setMessage(message);
        msg.setTime(TimeFactory.nowFactory.call());

        repo.add(msg);
        uow.saveChanges();
        uow.close();

        NotificationServiceFactory.instance.getService().notifyNewMessage(msg);
    }
}
