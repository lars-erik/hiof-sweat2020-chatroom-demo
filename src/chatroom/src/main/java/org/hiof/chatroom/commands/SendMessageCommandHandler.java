package org.hiof.chatroom.commands;

import org.hiof.chatroom.core.ChatMessage;
import org.hiof.chatroom.core.TimeFactory;
import org.hiof.chatroom.notification.NotificationService;
import org.hiof.chatroom.persistence.Repository;
import org.hiof.chatroom.persistence.UnitOfWork;

import java.util.UUID;

public class SendMessageCommandHandler implements CommandHandler {
    private Repository<ChatMessage> repository;
    private UnitOfWork unitOfWork;
    private NotificationService notificationService;

    public SendMessageCommandHandler(Repository<ChatMessage> repository, UnitOfWork unitOfWork, NotificationService notificationService) {
        this.repository = repository;
        this.unitOfWork = unitOfWork;
        this.notificationService = notificationService;
    }

    @Override
    public void execute(Command command) throws Exception {
        execute((SendMessageCommand) command);
    }

    private void execute(SendMessageCommand command) throws Exception {
        try {
            ChatMessage msg = new ChatMessage();
            msg.setId(UUID.randomUUID().toString());
            msg.setUser(command.user);
            msg.setMessage(command.message);
            msg.setTime(TimeFactory.nowFactory.call());

            repository.add(msg);
            unitOfWork.saveChanges();

            notificationService.notifyNewMessage(msg);
        }
        finally {
            unitOfWork.close();
        }
    }
}
