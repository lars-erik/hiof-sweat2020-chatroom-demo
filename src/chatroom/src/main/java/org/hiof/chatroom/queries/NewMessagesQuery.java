package org.hiof.chatroom.queries;

import org.hiof.chatroom.core.ChatMessage;
import org.hiof.chatroom.persistence.PersistenceFactory;
import org.hiof.chatroom.persistence.Repository;
import org.hiof.chatroom.persistence.UnitOfWork;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class NewMessagesQuery extends Query {
    public int limit;

    public NewMessagesQuery(int limit) {
        this.limit = limit;
    }

    public List<String> execute() {
        try {
            UnitOfWork unitOfWork = PersistenceFactory.Instance.createUnitOfWork();
            Repository<ChatMessage> chatMessageRepository = PersistenceFactory.Instance.createChatMessageRepository(unitOfWork);
            List<String> lastMessages = chatMessageRepository.query(this).map(x -> x.toString()).collect(Collectors.toList());
            unitOfWork.close();
            return lastMessages;
        } catch (Exception e) {
            return new ArrayList<>();
        }
    }
}
