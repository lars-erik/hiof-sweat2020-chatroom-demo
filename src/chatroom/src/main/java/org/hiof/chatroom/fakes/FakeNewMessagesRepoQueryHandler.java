package org.hiof.chatroom.fakes;

import org.hiof.chatroom.core.ChatMessage;
import org.hiof.chatroom.queries.Query;
import org.hiof.chatroom.persistence.RepositoryQueryHandler;
import org.hiof.chatroom.persistence.Repository;

import java.util.stream.Stream;

public class FakeNewMessagesRepoQueryHandler implements RepositoryQueryHandler<ChatMessage> {
    @Override
    public Stream<ChatMessage> query(Query query, Repository repository) {
        FakeChatMessageRepository repo = (FakeChatMessageRepository)repository;

        long count = repo.messages.stream().count();
        Stream<ChatMessage> sorted = repo.messages.stream()
                .sorted((x, y) -> Long.compare(y.getTime().getEpochSecond(), x.getTime().getEpochSecond()))
                .limit(20);
        return sorted;
    }
}
