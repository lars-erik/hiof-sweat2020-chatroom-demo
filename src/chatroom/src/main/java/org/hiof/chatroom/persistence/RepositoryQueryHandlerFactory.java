package org.hiof.chatroom.persistence;

import org.hiof.chatroom.queries.Query;

import java.util.HashMap;
import java.util.Map;

public class RepositoryQueryHandlerFactory {
    static Map<Class<? extends Query>, Class<? extends RepositoryQueryHandler>> handlers = new HashMap<>();

    public static void register(Class<? extends Query> queryType, Class<? extends RepositoryQueryHandler> handlerType)
    {
        handlers.put(queryType, handlerType);
    }

    public static RepositoryQueryHandler createFor(Query query) throws Exception
    {
        Class<? extends RepositoryQueryHandler> handlerType = handlers.get(query.getClass());
        if (handlerType == null) {
            throw new Exception("No registered handler, or not parameterless constructor");
        }
        try {
            RepositoryQueryHandler handler = handlerType.newInstance();
            return handler;
        } catch (InstantiationException e) {
            throw new Exception("No registered handler, or not parameterless constructor");
        } catch (IllegalAccessException e) {
            throw new Exception("No registered handler, or not parameterless constructor");
        }
    }
}
