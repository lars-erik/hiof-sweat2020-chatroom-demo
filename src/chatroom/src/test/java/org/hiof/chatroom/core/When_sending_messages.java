package org.hiof.chatroom.core;

import org.hiof.chatroom.commands.SendMessageCommand;
import org.hiof.chatroom.commands.SendMessageCommandHandler;
import org.hiof.chatroom.fakes.FakeNewMessagesRepoQueryHandler;
import org.hiof.chatroom.notification.NotificationService;
import org.hiof.chatroom.notification.NotificationServiceFactory;
import org.hiof.chatroom.persistence.Repository;
import org.hiof.chatroom.persistence.RepositoryQueryHandlerFactory;
import org.hiof.chatroom.persistence.UnitOfWork;
import org.hiof.chatroom.queries.NewMessagesQuery;
import org.hiof.chatroom.support.PersistenceSupport;
import org.junit.jupiter.api.*;
import org.mockito.Mockito;

import java.time.Instant;

public class When_sending_messages {
    protected PersistenceSupport persistenceSupport;
    protected NotificationService notificationService;

    @BeforeEach
    public void initialize_persistence() throws Exception {
        persistenceSupport = new PersistenceSupport();
    }

    @BeforeEach
    public void initialize_notification() {
        notificationService = Mockito.mock(NotificationService.class);
    }

    @AfterEach
    public void reset_time() {
        TimeFactory.reset();
    }

    @AfterEach
    public void cleanup_database() throws Exception {
        persistenceSupport.cleanup();
    }

    @Test
    public void message_are_stored_in_database() throws Exception {
        SendMessageCommand cmd = sendMessage();
        ChatMessage last = getLastMessage();

        Assertions.assertEquals(cmd.user, last.getUser());
        Assertions.assertEquals(cmd.message, last.getMessage());
    }

    @Test
    public void message_are_stamped_with_time() throws Exception {
        TimeFactory.nowFactory = () -> Instant.parse("2020-10-01T23:59:59Z");

        SendMessageCommand cmd = sendMessage();

        ChatMessage last = getLastMessage();

        Assertions.assertEquals("2020-10-01T23:59:59Z", last.getTime().toString());
    }

    @Test
    public void notification_is_sent_to_all_users() throws Exception {
        SendMessageCommand cmd = sendMessage();
        ChatMessage last = getLastMessage();
        String expected = last.toString();

        Mockito.verify(notificationService).notifyNewMessage(Mockito.argThat(x -> expected.equals(x.toString())));
    }

    private ChatMessage getLastMessage() throws Exception {
        ChatMessage last = null;
        last = persistenceSupport.getChatMessageRepository().query(new NewMessagesQuery(1)).findFirst().get();
        return last;
    }

    protected SendMessageCommand sendMessage() throws Exception {
        SendMessageCommand cmd = new SendMessageCommand();
        cmd.user = "Luke Skywalker";
        cmd.message = "This is Red Leader. We're approaching the Ison Corridor!";

        SendMessageCommandHandler handler = new SendMessageCommandHandler(
            persistenceSupport.getChatMessageRepository(),
            persistenceSupport.getUnitOfWork(),
            notificationService
        );
        handler.execute(cmd);

        return cmd;
    }
}
