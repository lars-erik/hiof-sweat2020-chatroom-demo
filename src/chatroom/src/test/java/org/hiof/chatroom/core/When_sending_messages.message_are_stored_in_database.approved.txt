ChatMessage{time='2020-10-01T23:59:59Z', user='Luke Skywalker', message='This is Red Leader. We're approaching the Ison Corridor!'}